In [21]:
#Bibek Shiwakoti 
# feb 16 2024

In [22]:
import numpy as np
import cv2

In [23]:
img = cv2.imread('./images/faces.jpg')
cv2.imshow('faces',img)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [24]:
face_detection_model =cv2.dnn.readNetFromCaffe('./model/deploy.prototxt',
                                                './model/res10_300x300_ssd_iter_140000.caffemodel')

In [41]:
def face_detection_dnn(img):
    blob = cv2.dnn.blobFromImage(img,1,(300,300),(104,177,123),swapRB=True)
    face_detection_model.setInput(blob)
    detections=face_detection_model.forward()
    image =img.copy()
    h,w =image.shape[:2]
    
    for i in range (0,detections.shape[2]):
        confidence =detections[0,0,i,2]
        if confidence >0.5:
            box=detections[0,0,i,3:7]*np.array([w,h,w,h])
            box=box.astype('int')
            pt1=(box[0],box[1])
            pt2=(box[2],box[3])
            
            cv2.reactangle(image,pt1,pt2,(0,255,0),1)
            
            text ='score : {:.0f} %'.format(confidence*100)
            cv2.putText(image,text,pt1,cv2,FONT_HERSHEY_SIMPLEX,1,(255,255,255),2)
            
            return image
            

In [42]:
cap=cv2.VideoCapture(0)
while True:
    ret,frame=cap.read()
    
    if ret == False:
        break
    img_detection = face_detection_dnn(frame)
    cv2.imshow('Real time face detection with dnn',img_detection)
    if cv2.waitKey(1)== ord('q'):
        break
        
cap.replace()
cv2.destroyAllWindows()

error: OpenCV(4.9.0) D:\a\opencv-python\opencv-python\opencv\modules\dnn\src\layers\convolution_layer.cpp:368: error: (-215:Assertion failed) !blobs.empty() || inputs.size() > 1 in function 'cv::dnn::ConvolutionLayerImpl::getMemoryShapes'


In [35]:
face_detection_model =cv2.dnn.readNetFromCaffe('./model/deploy.prototxt',
                                                './model/New Folder/res10_300x300_ssd_iter_140000.caffemodel')

In [36]:
# step1: blob from image
blob = cv2.dnn.blobFromImage(img,1,(300,300),(104,177,123),swapRB=True)
# step2 : set blob as input
face_detection_model.setInput(blob)
#step 3: get the output
detections = face_detection_model.forward()


error: OpenCV(4.9.0) D:\a\opencv-python\opencv-python\opencv\modules\dnn\src\layers\convolution_layer.cpp:368: error: (-215:Assertion failed) !blobs.empty() || inputs.size() > 1 in function 'cv::dnn::ConvolutionLayerImpl::getMemoryShapes'


In [ ]:
# ***********************************

error                                     Traceback (most recent call last)
Cell In[36], line 6
      4 face_detection_model.setInput(blob)
      5 #step 3: get the output
----> 6 detections = face_detection_model.forward()

error: OpenCV(4.9.0) D:\a\opencv-python\opencv-python\opencv\modules\dnn\src\layers\convolution_layer.cpp:368: error: (-215:Assertion failed) !blobs.empty() || inputs.size() > 1 in function 'cv::dnn::ConvolutionLayerImpl::getMemoryShapes'


In [ ]:
# *******************tried to debug and resolve blob empty error but cant resolve it right now**********************************************

In [7]:
import cv2
import numpy as np

# Load the image
img = cv2.imread('./images/faces.jpg')

# Check if the image was loaded successfully
if img is None:
    print("Error: Failed to load the image.")
else:
    # Load the face detection model
    face_detection_model = cv2.dnn.readNetFromCaffe('./model/deploy.prototxt',
                                                    './model/res10_300x300_ssd_iter_140000.caffemodel')

    # Check if the model was loaded successfully
    if face_detection_model.empty():
        print("Error: Failed to load the face detection model.")
    else:
        print("Model loaded successfully.")
        
        # Create a blob from the image
        blob = cv2.dnn.blobFromImage(img, 1.0, (300, 300), (104.0, 177.0, 123.0))

        # Check if the blob was created successfully
        if blob is None:
            print("Error: Failed to create the blob.")
        else:
            print("Blob created successfully.")
            
            # Set the blob as input to the model
            face_detection_model.setInput(blob)

            # Perform inference
            detections = face_detection_model.forward()

            # Check if detections were successful
            if detections is None:
                print("Error: No detections found.")
            else:
                print("Detections:", detections)


Model loaded successfully.
Blob created successfully.


error: OpenCV(4.9.0) D:\a\opencv-python\opencv-python\opencv\modules\dnn\src\layers\convolution_layer.cpp:368: error: (-215:Assertion failed) !blobs.empty() || inputs.size() > 1 in function 'cv::dnn::ConvolutionLayerImpl::getMemoryShapes'


In [9]:
import cv2
import numpy as np

# Load the image
img = cv2.imread('./images/faces.jpg')

# Check if the image was loaded successfully
if img is None:
    print("Error: Failed to load the image.")
else:
    # Load the face detection model
    face_detection_model = cv2.dnn.readNetFromCaffe('./model/deploy.prototxt',
                                                    './model/res10_300x300_ssd_iter_140000.caffemodel')

    # Check if the model was loaded successfully
    if face_detection_model.empty():
        print("Error: Failed to load the face detection model.")
    else:
        print("Model loaded successfully.")
        
        # Print the names of the input layers
        print("Input layer names:", face_detection_model.getLayerNames())
        
        # Create a blob from the image
        blob = cv2.dnn.blobFromImage(img, 1.0, (300, 300), (104.0, 177.0, 123.0))

        # Check if the blob was created successfully
        if blob is None:
            print("Error: Failed to create the blob.")
        else:
            print("Blob created successfully.")
            
            # Set the blob as input to the model
            input_layer_name = face_detection_model.getLayerNames()[0]
            face_detection_model.setInput(blob, input_layer_name)
    

            # Perform inference
            detections = face_detection_model.forward()

            # Check if detections were successful
            if detections is None:
                print("Error: No detections found.")
            else:
                print("Detections:", detections)


Model loaded successfully.
Input layer names: ('conv1', 'relu1', 'pool1', 'norm1', 'conv2', 'relu2', 'pool2', 'norm2', 'conv3', 'relu3', 'conv4', 'relu4', 'conv5', 'relu5', 'pool5', 'fc6', 'relu6', 'drop6', 'fc7', 'relu7', 'drop7', 'fc8', 'prob')
Blob created successfully.


error: OpenCV(4.9.0) D:\a\opencv-python\opencv-python\opencv\modules\dnn\src\net_impl.cpp:1403: error: (-204:Requested object was not found) Requested blob "conv1" not found in function 'cv::dnn::dnn4_v20231225::Net::Impl::setInput'
